text/0249-v2-experiments.md

```diff
@@ -408,3 +484,240 @@ Disadvantages:
+matrix:
+
+| .                | Stable App        | Experimental App         |
+|------------------|-------------------|--------------------------|
+| Stable Lib       | floating versions | lib floating, app pinned |
+| Experimental Lib | -                 | pinned versions          |
+
+> NOTE: in the following sections, I'll be using the terms "stable app", "experimental app",
+> "stable lib" and "experimental lib" as stand-ins for the more accurate but much
+> longer "app that depends on stable CDK", "app that depends on experimental CDK",
+> etc.
+
+## Different package names
+
+We would vend two different package names, for example:
+
+- `aws-cdk-lib`
+- `aws-cdk-lib-experimental`
+
+The tl;dr of this section is as follows. Read below for details.
+
+| .     | Stable apps | Experimental app uses Stable lib | Without complaints | Lib advertises Experimental |
+|-------|-------------|----------------------------------|--------------------|-----------------------------|
+| NPM   | yes         | with package aliases             | yes                | no                          |
+| Maven | yes         | with excluded dependencies       | yes                | yes                         |
+| pip   | yes         | not possible                     | -                  | no                          |
+| NuGet | ?           | ?                                | ?                  | ?                           |
+| Go    | ?           | ?                                |                    | ?                           |
+
+**NPM**
+
+In NPM-based languages (JavaScript and TypeScript), the package name appears
+in source code (in the `require()` statement). If we changed the library
+name, we wouldn't be able to use a library that uses stable code (it would
+contain `require('aws-cdk-lib')`) in an application that uses experimental
+code (the `package.json` would contain: `{ dependencies: {
+"aws-cdk-lib-experiments": "..." }}`).
+
+NPM 6.9.0 (released May 2019) introduces a feature that helps with this:
+[package
+aliases](https://github.com/npm/rfcs/blob/latest/implemented/0001-package-aliases.md)
+which allows installing a package under a different name, so we could install
+`aws-cdk-lib-experimental` under the name `aws-cdk-lib`, and no `require()` statements
+would have to be changed.
+
+An application trying to use experimental will do this:
+
+```json
+{
+  "dependencies": {
+      "aws-cdk-lib": "npm:aws-cdk-lib-experiments@1.60.0",
+      "some-construct-lib-that-uses-stable": "^1.2.3"
+  }
+}
+```
+
+An experimental library would declare this:
+
+```json
+{
+  "peerDependencies": {
+      "aws-cdk-lib": "1.60.0"
+      // Note: ideally you'd like to express "aws-cdk-lib-experiments": "1.60.0" here, but
+      // if you do that "npm ls" will complain about an unmet peer dependency
+  },
+  "devDependencies": {
+      "aws-cdk-lib": "npm:aws-cdk-lib-experiments@1.60.0"
+  }
+}
+```
```

Comment:
Sorry, I don't understand the problem here. If a library peer depends on `aws-cdk-lib-experiments`, then the application _should_ declare a dependency on `aws-cdk-lib-experiments`, and NPM complaining about missing peer dependency if the application only depends on `aws-cdk-lib` is a _good_ thing I'm pretty sure?

What am I missing?

Reply:
I see I left something out that was (to me) so obvious that I didn't think it would need writing down. Failure of judgement :).